src/transform/src/dataflow.rs

```diff
@@ -434,76 +448,382 @@ pub fn optimize_dataflow_monotonic(dataflow: &mut DataflowDesc) -> Result<(), Tr
+            // - Some indexes might be less skewed than others.
+            // - Some indexes might have an error, while others don't.
+            //   https://github.com/MaterializeInc/materialize/issues/15557
+            // - Some indexes might have less extra data in their keys, which won't be used in a
+            //   full scan.
             if let Some(key) = indexes.indexes_on(*id).next() {
-                keys.insert(key.to_owned());
+                requests_with_matching_indexes.push((key.to_owned(), IndexUsageType::FullScan));
             }
         }
     }
-    dataflow.index_imports.retain(|_, (index_desc, _, _)| {
-        if let Some(keys) = indexes_by_view.get(&index_desc.on_id) {
-            keys.contains(&index_desc.key)
-        } else {
-            false
+
+    // Annotate index imports by their usage types
+    for (
+        _index_id,
+        IndexImport {
+            desc: index_desc,
+            typ: _,
+            monotonic: _,
+            usage_types,
+        },
+    ) in dataflow.index_imports.iter_mut()
+    {
+        if let Some(requested_accesses) = index_reqs_by_id.get(&index_desc.on_id) {
+            assert!(usage_types.is_none()); // We should be called on a dataflow only once
+            *usage_types = Some(Vec::new());
+            assert!(!requested_accesses.is_empty()); // See above at the "Add full index scans"
+            for (req_key, req_usage_type) in requested_accesses {
+                if *req_key == index_desc.key {
+                    usage_types.as_mut().unwrap().push(req_usage_type.clone());
+                }
+            }
         }
-    });
+    }
+
+    // Prune index imports to only those that are used
+    dataflow
+        .index_imports
+        .retain(|_, index_import| match &index_import.usage_types {
+            None => false,
+            Some(usage_types) => {
+                !usage_types.is_empty()
+                // (can be empty here when an other index on this same object is used)
+            }
+        });
 
     mz_repr::explain::trace_plan(dataflow);
 
     Ok(())
 }
+
+#[derive(Debug)]
+struct CollectIndexRequests<'a> {
+    // We were told about these indexes being available.
+    indexes_available: &'a dyn IndexOracle,
+    // We'll be collecting index requests here.
+    index_reqs_by_id: &'a mut BTreeMap<GlobalId, Vec<(Vec<MirScalarExpr>, IndexUsageType)>>,
+    // As we recurse down a MirRelationExpr, we'll need to keep track of the context of the
+    // current node (see docs on `IndexUsageContext` about what context we keep).
+    // Moreover, we need to propagate this context from cte uses to cte definitions.
+    // `context_across_lets` will keep track of the contexts that reached each use of a LocalId
+    // added together.
+    context_across_lets: BTreeMap<LocalId, Vec<IndexUsageContext>>,
+    recursion_guard: RecursionGuard,
+}
+
+impl<'a> CheckedRecursion for CollectIndexRequests<'a> {
```

Comment:
I don't think you are relying on the `CheckedRecursion` implementation anywhere. I was able to remove it together with the `recursion_guard: RecursionGuard` field and the code still typechecks.

You need to wrap the body of `collect_index_reqs_inner` in a `checked_recur` call.

```rust
self.checked_recur(|_| {
    // body
}
```

Alternatively, you do

```rust
self.recursion_guard.checked_recur(|_| {
    // body
}
```

and then you don't need to implement `CheckedRecursion` for `CollectIndexRequests`.

Reply:
Fixed!